File: playground/Container With Most Water.js
/**
 * @param {number[]} height
 * @return {number}
 */


/*
    double pointer solution 

    the water volume =  weight (right - left) * height ( min( nums[left] , nums[right] )   ) 
    from outside to inside 
*/

var maxArea = function(height) {

    let left = 0 ; 
    let right = height.length - 1 ; 

    let best = Number.MIN_SAFE_INTEGER  ; 

    // if left == right : volume => 0 
    while ( left < right ){

        const volume = (right-left) * Math.min(  height[left] , height[right]    ) ; 
        best = Math.max(best , volume) ; 

        // update the pointer : 
        if (height[left] > height[right]) {
            right -= 1 ; 
        }
        else {
            left += 1 ; 
        }
    }

    return best ; 
    
};

File: TOP150Review/Container With Most Water.py

from typing import List 

class Solution:
    
    # [ 1,6,3,5,2 ]
    def maxArea(self, height: List[int]) -> int:
    
        left = 0 
        right = len(height) - 1  
        best = float("-inf")
        # width = right - left   
        while right > left : 
            
            if height[left] > height[right] : 
                best = max(best , (right-left)*height[right]  )
                right -=  1 
            else : 
                best = max(best , (right-left)*height[left]  )
                left += 1 
        
        return best 
            
            